### PR TITLE
Add page allocator

### PR DESCRIPTION
**Pull request by Skallwar**

### CODE REVIEW

File: .github/workflows/kunit.yml [hunk 1]
```diff
@@ -49,4 +49,4 @@ jobs:
         uses: actions-rs/cargo@v1
         with:
           command: test
-          args: --target ${{ matrix.target }} --verbose
+          args: --target ${{ matrix.target }} --release --verbose
```
Comment (CohenArthur) [inline .github/workflows/kunit.yml:52]: Why release here?

File: src/allocator/mod.rs [hunk 2]
```diff
@@ -0,0 +1,93 @@
+// use crate::arch;
+extern crate spin;
```
Comment (CohenArthur) [inline src/allocator/mod.rs:2]: We don't need to mark extern crates as extern anymore, but I'm not against it if it is for clarity

File: src/allocator/mod.rs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        let heap_pages_count: usize = (heap_end - heap_start) / arch::PAGE_SIZE;
+        let metadata_overhead: usize = (heap_pages_count) * mem::size_of::<UsageFlags>();
+        // Calculat the address of the first page after heap_start + metadata
+        let first_allocatable_page: usize =
+            ((heap_start + metadata_overhead + arch::PAGE_SIZE - 1) / arch::PAGE_SIZE)
+                * arch::PAGE_SIZE;
+        let allocatable_pages_count =
+            heap_pages_count - (first_allocatable_page - heap_start) / arch::PAGE_SIZE;
+
+        // println!("heap_start = {}\nheap_end = {}\nheap_pages_count = {}\nmetadata_overhead = {}\nfirst_allocatable_page = {:#x}\nallocatable_pages_count= {}", heap_start, heap_end, heap_pages_count, metadata_overhead, first_allocatable_page, allocatable_pages_count);
```
Comment (CohenArthur) [inline src/allocator/mod.rs:43]: Dead code?

File: src/allocator/mod.rs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    #[test_case]
+    fn page_alloc_out_of_memory() {
+        let mut allocator = PAGE_ALLOC.write();
+        while allocator.page_alloc().is_some() {
+            // At some point we will not have any free pages left
+        }
+
+        kassert_eq!(true, true, "Page alloc out of memory");
```
Comment (CohenArthur) [inline src/allocator/mod.rs:91]:
```suggestion
        kassert!(true, "Page alloc out of memory");
```

File: src/allocator/mod.rs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    }
```
Comment (CohenArthur) [inline src/allocator/mod.rs:92]: Maybe add a test to make sure that the next page we get is `None`

File: src/allocator/mod.rs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+impl PageAllocator {
+    pub fn new() -> PageAllocator {
+        let heap_start: usize = unsafe { &arch::HEAP_START as *const () } as usize;
+        let heap_end: usize = unsafe { &arch::HEAP_END as *const () } as usize;
+
+        let heap_pages_count: usize = (heap_end - heap_start) / arch::PAGE_SIZE;
+        let metadata_overhead: usize = (heap_pages_count) * mem::size_of::<UsageFlags>();
+        // Calculat the address of the first page after heap_start + metadata
```
Comment (CohenArthur) [inline src/allocator/mod.rs:36]:
```suggestion
        // Calculate the address of the first page after heap_start + metadata
```

File: src/allocator/mod.rs [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+pub struct PageAllocator {
+    usage: &'static mut [UsageFlags],
+    number_of_pages: usize,
+    first_page: usize,
+}
+
+impl PageAllocator {
+    pub fn new() -> PageAllocator {
+        let heap_start: usize = unsafe { &arch::HEAP_START as *const () } as usize;
+        let heap_end: usize = unsafe { &arch::HEAP_END as *const () } as usize;
```
Comment (CohenArthur) [inline src/allocator/mod.rs:32]: We should definitely add some sort of macro or function for this, I feel like it's quite useful and that we're using it quite often

File: src/allocator/mod.rs [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::utest::uassert_eq;
+
+    #[test_case]
+    fn page_alloc_one() {
+        let page = PAGE_ALLOC.write().page_alloc();
+        kassert_eq!(page.is_some(), true, "Page alloc one page test");
```
Comment (CohenArthur) [inline src/allocator/mod.rs:81]:
```suggestion
        kassert!(page.is_some(), "Page alloc one page test");
```

**Review by CohenArthur — COMMENTED** [review]
Looks super nice. From what I understand, the layout is the following, right? 

---- HEAP START -----
---- Array of metadata (free, used) -----
---- First page ----
...
---- HEAP END ----

I feel like the module overall could use some more documentation, but we can merge and fix it later. Thanks a lot :fire: :fire:

**Comment by Skallwar** [comment]
> Looks super nice. From what I understand, the layout is the following, right?
> 
> ---- HEAP START -----
> ---- Array of metadata (free, used) -----
> ---- First page ----
> ...
> ---- HEAP END ----

Exactly

> I feel like the module overall could use some more documentation, but we can merge and fix it later.

I agree. That the first thing I (or we) will do after we define the ROADMAP. I just don't want to document code that might be trashed away after our meeting :sweat_smile:
We need to document everything that has been done

File: .github/workflows/kunit.yml (hunk 1 above)
Comment (Skallwar) [reply .github/workflows/kunit.yml:52]: Because ``page_alloc_out_of_memory`` test is really slow without it (more than a minute)

File: src/allocator/mod.rs (hunk 2 above)
Comment (Skallwar) [reply src/allocator/mod.rs:2]: We can just
 ``` rust
use spin::....
```
?

File: src/allocator/mod.rs (hunk 7 above)
Comment (Skallwar) [reply src/allocator/mod.rs:32]: Yes I agree. Macro or function.

File: src/allocator/mod.rs (hunk 5 above)
Comment (Skallwar) [reply src/allocator/mod.rs:92]: By definition it is since we are now out of the loop

File: src/allocator/mod.rs (hunk 2 above)
Comment (CohenArthur) [reply src/allocator/mod.rs:2]: Yep, extern is not needed anymore since Rust 2018

File: src/allocator/mod.rs (hunk 5 above)
Comment (CohenArthur) [reply src/allocator/mod.rs:92]: That's fair

File: src/allocator/mod.rs (hunk 2 above)
Comment (Skallwar) [reply src/allocator/mod.rs:2]: Oups

**Comment by Skallwar** [comment]
We will use an already existing crate
